engine/hatchery/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,307 @@
+	return h.client
+}
+
+//Configuration returns Hatchery CommonConfiguration
+func (h *HatcheryKubernetes) Configuration() hatchery.CommonConfiguration {
+	return h.Config.CommonConfiguration
+}
+
+// ModelType returns type of hatchery
+func (*HatcheryKubernetes) ModelType() string {
+	return sdk.Kubernetes
+}
+
+// CanSpawn return wether or not hatchery can spawn model.
+// requirements are not supported
+func (h *HatcheryKubernetes) CanSpawn(model *sdk.Model, jobID int64, requirements []sdk.Requirement) bool {
+	for _, r := range requirements {
+		if r.Type == sdk.ServiceRequirement {
+			return false
+		}
+	}
+	return true
+}
+
+// SpawnWorker starts a new worker process
+func (h *HatcheryKubernetes) SpawnWorker(spawnArgs hatchery.SpawnArguments) (string, error) {
+	name := fmt.Sprintf("k8s-%s-%s", strings.ToLower(spawnArgs.Model.Name), strings.Replace(namesgenerator.GetRandomName(0), "_", "-", -1))
+	label := "execution"
+	if spawnArgs.RegisterOnly {
+		name = "register-" + name
+		label = "register"
+	}
+
+	envs := []apiv1.EnvVar{
+		{Name: "CDS_API", Value: h.Config.API.HTTP.URL},
+		{Name: "CDS_NAME", Value: name},
+		{Name: "CDS_TOKEN", Value: h.Configuration().API.Token},
+		{Name: "CDS_SINGLE_USE", Value: "1"},
+		{Name: "CDS_MODEL", Value: fmt.Sprintf("%d", spawnArgs.Model.ID)},
+		{Name: "CDS_HATCHERY", Value: fmt.Sprintf("%d", h.hatch.ID)},
+		{Name: "CDS_HATCHERY_NAME", Value: h.hatch.Name},
+		{Name: "CDS_FORCE_EXIT", Value: "1"},
+		{Name: "CDS_TTL", Value: fmt.Sprintf("%d", h.Config.WorkerTTL)},
+	}
+
+	if spawnArgs.JobID > 0 {
+		if spawnArgs.IsWorkflowJob {
+			envs = append(envs, apiv1.EnvVar{Name: "CDS_BOOKED_WORKFLOW_JOB_ID", Value: fmt.Sprintf("%d", spawnArgs.JobID)})
+		} else {
+			envs = append(envs, apiv1.EnvVar{Name: "CDS_BOOKED_PB_JOB_ID", Value: fmt.Sprintf("%d", spawnArgs.JobID)})
+		}
+	}
+
+	var gracePeriodSecs int64
+	pod, err := h.k8sClient.CoreV1().Pods(h.Config.KubernetesNamespace).Create(&apiv1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			DeletionGracePeriodSeconds: &gracePeriodSecs,
+			Labels: map[string]string{
+				LABEL_WORKER:       label,
+				LABEL_WORKER_MODEL: strings.ToLower(spawnArgs.Model.Name),
+			},
+		},
+		Spec: apiv1.PodSpec{
+			RestartPolicy:                 apiv1.RestartPolicyNever,
+			TerminationGracePeriodSeconds: &gracePeriodSecs,
+			Containers: []apiv1.Container{
+				{
+					Name:  name,
+					Image: spawnArgs.Model.Image,
+					Env:   envs,
+					Resources: apiv1.ResourceRequirements{
+						Requests: apiv1.ResourceList{
+							apiv1.ResourceMemory: resource.MustParse(fmt.Sprintf("%d", h.Config.DefaultMemory)),
+						},
+					},
+				},
+			},
+		},
+	})
+
+	return pod.Name, err
+}
+
+// WorkersStarted returns the number of instances started but
+// not necessarily register on CDS yet
+func (h *HatcheryKubernetes) WorkersStarted() int {
+	list, err := h.k8sClient.CoreV1().Pods(h.Config.KubernetesNamespace).List(metav1.ListOptions{LabelSelector: LABEL_WORKER})
```

Comment:
should we list by the hatchery name (and set the hatchery name as a label) ?

Reply:
Yes I can add another label for hatchery name